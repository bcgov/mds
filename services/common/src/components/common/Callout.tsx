import React, { FC, ReactNode } from "react";

import { CALLOUT_SEVERITY } from "@mds/common/constants/strings";

interface CallOutProps {
  message: string | ReactNode;
  title?: string;
  severity: string;
}

const Callout: FC<CallOutProps> = ({ message, title, severity = CALLOUT_SEVERITY.info }) => {
  const formattedMessage = typeof message === "string" ? <p>{message}</p> : message;

  return (
    <div className={`bcgov-callout--${severity} bcgov-callout`}>
      {title && <p className="bcgov-callout-title">{title}</p>}
      {formattedMessage}
    </div>
  );
};

export default Callout;
