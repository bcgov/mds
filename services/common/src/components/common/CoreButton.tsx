import React, { FC } from "react";
import { Button, ButtonProps } from "antd";
import { ButtonType } from "antd/lib/button";

// DO NOT put antd button types in here: default, primary, ghost, dashed, link, text
declare const AdditionalTypes: ["tertiary", "filled-tertiary"];
// Additional doesn't exist at compile-time, I have not found a way around this. :(
// for now the two arrays should be duplicated.
const additionalTypes = ["tertiary", "filled-tertiary"];

type CoreCustomType = typeof AdditionalTypes[number];

// this one is the default because it's not used anywhere in the system,
// therefore it can be styled freely
const defaultButtonType = "dashed";

interface CoreButtonProps extends Omit<ButtonProps, "type"> {
  type: ButtonType | CoreCustomType;
}

const CoreButton: FC<CoreButtonProps> = ({ type, className, children, ...props }) => {
  const isAntdType = additionalTypes.includes(type);
  const buttonType = isAntdType ? defaultButtonType : (type as ButtonType);
  const buttonClassName = ["core-btn", `core-btn-${type}`, className].join(" ").trim();

  return (
    <Button {...props} type={buttonType} className={buttonClassName}>
      {children}
    </Button>
  );
};

export default CoreButton;
