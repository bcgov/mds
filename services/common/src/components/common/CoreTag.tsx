import React, { FC, ReactNode } from "react";
import { Row, Typography } from "antd";
import { Link } from "react-router-dom";

interface TagProps {
  text: string;
  icon: ReactNode;
  link?: string;
}

const CoreTag: FC<TagProps> = ({ text, icon, link }) => {
  const getText = () => {
    return link ? (
      <Link style={{ textDecoration: "none", color: "inherit" }} to={link}>
        {text}
      </Link>
    ) : (
      text
    );
  };

  return (
    <Row justify="space-between" align="middle" className="tag">
      {icon}
      <Typography.Text className="margin-medium--left">{getText()}</Typography.Text>
    </Row>
  );
};

export default CoreTag;
