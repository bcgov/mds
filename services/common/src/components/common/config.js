import RenderSelect from "../forms/RenderSelect";
import RenderDate from "./RenderDate";
import RenderYear from "./RenderYear";

// This file is anticipated to have multiple exports
// eslint-disable-next-line import/prefer-default-export
export const renderConfig = {
  SELECT: RenderSelect,
  DATE: RenderDate,
  YEAR: RenderYear,
};
