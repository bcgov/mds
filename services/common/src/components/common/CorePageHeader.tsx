import React, { FC, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Col, Row, Tabs, TabsProps, Typography } from "antd";
import { Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import CoreTag from "./CoreTag";

import CompanyIcon from "@mds/common/assets/icons/CompanyIcon";
import { faLocationDot } from "@fortawesome/pro-light-svg-icons";
import { getMineById } from "@mds/common/redux/selectors/mineSelectors";
import { fetchMineRecordById } from "@mds/common/redux/actionCreators/mineActionCreator";

interface BreadCrumb {
  route: string;
  text: string;
}

interface CorePageHeaderProps {
  entityLabel: string;
  entityType: string;
  mineGuid: string;
  current_permittee: string; // would be ideal to get this from the mine
  breadCrumbs?: BreadCrumb[];
  tabProps?: TabsProps;
}

const { Title, Text } = Typography;

const CorePageHeader: FC<CorePageHeaderProps> = ({
  mineGuid,
  current_permittee,
  entityLabel,
  entityType,
  breadCrumbs,
  tabProps,
}) => {
  const mine = useSelector((state) => getMineById(state, mineGuid));
  const dispatch = useDispatch();

  useEffect(() => {
    if (!mine) {
      dispatch(fetchMineRecordById(mineGuid));
    }
  }, [mineGuid]);

  return (
    <div className="core-page">
      <div className="view--header padding-lg--top padding-lg--sides core-page-header">
        <Row>
          <Col>
            {breadCrumbs.map((crumb) => {
              return (
                <React.Fragment key={crumb.route}>
                  <Link to={crumb.route} className="faded-text">
                    {crumb.text}
                  </Link>{" "}
                  /{" "}
                </React.Fragment>
              );
            })}
            <Text>
              {entityType} {entityLabel}
            </Text>
          </Col>
        </Row>
        <Row align="middle" gutter={16}>
          <Col>
            <Title level={1} className="margin-none">
              {entityType} {entityLabel}
            </Title>
          </Col>
          <Col>
            <CoreTag
              icon={<FontAwesomeIcon icon={faLocationDot} />}
              text={mine?.mine_name}
              link={GLOBAL_ROUTES?.MINE_DASHBOARD.dynamicRoute(mineGuid)}
            />
          </Col>
          <Col>
            <CoreTag icon={<CompanyIcon />} text={current_permittee} />
          </Col>
        </Row>
      </div>
      {tabProps && (
        <Tabs
          className="core-tabs fixed-tabs-tabs padding-md--top"
          {...tabProps}
          tabBarStyle={{ paddingLeft: 20, paddingRight: 20 }}
        />
      )}
    </div>
  );
};

export default CorePageHeader;
