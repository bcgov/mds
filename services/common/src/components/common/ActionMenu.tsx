import React, { FC, ReactNode } from "react";
import { Button, Dropdown, Modal } from "antd";
import CaretDownOutlined from "@ant-design/icons/CaretDownOutlined";
import DownOutlined from "@ant-design/icons/DownOutlined";
import { ITableAction } from "@mds/common/components/common/CoreTableCommonColumns";

export const deleteConfirmWrapper = (recordDescription: string, onOk: () => void) => {
  const title = `Confirm Deletion`;
  const content = `Are you sure you want to delete this ${recordDescription}?`;
  const modalContent = {
    title,
    content,
    onOk,
    okText: "Delete",
  };
  return Modal.confirm(modalContent);
};

export const generateActionMenuItems = (actionItems: ITableAction[], record) => {
  return actionItems.map((action) => {
    return {
      key: action.key,
      icon: action.icon,
      label: (
        <button
          type="button"
          disabled={action.disabled}
          className={`full actions-dropdown-button`}
          data-testid={`action-button-${action.key}`}
          onClick={(event) => action.clickFunction(event, record)}
        >
          <div>{action.label}</div>
        </button>
      ),
    };
  });
};

export interface IHeaderAction {
  key: string;
  label: string;
  icon?: ReactNode;
  clickFunction: () => void | Promise<void>;
}
// Looks like a button, intended for page-scope, not record-scope in the actions
export const ActionMenuButton: FC<{ buttonText?: string; actions: IHeaderAction[] }> = ({
  actions,
  buttonText = "Action",
}) => {
  const items = generateActionMenuItems((actions as unknown) as ITableAction[], null);

  return (
    <Dropdown menu={{ items }} placement="bottomLeft">
      <Button type="ghost" className="actions-dropdown-button">
        {buttonText}
        <DownOutlined />
      </Button>
    </Dropdown>
  );
};

interface ActionMenuProps {
  record: any;
  actionItems: ITableAction[];
  category: string;
}
const ActionMenu: FC<ActionMenuProps> = ({ record, actionItems, category }) => {
  const items = generateActionMenuItems(actionItems, record);
  return (
    <Dropdown menu={{ items }} placement="bottomLeft">
      <Button type="text" className="actions-dropdown-button">
        Actions
        <CaretDownOutlined alt={`${category} Actions`} />
      </Button>
    </Dropdown>
  );
};

export default ActionMenu;
