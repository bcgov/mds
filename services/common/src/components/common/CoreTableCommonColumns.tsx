import React, { ReactNode } from "react";
import Highlight from "react-highlighter";
import { dateSorter, formatDate, nullableStringSorter } from "@mds/common/redux/utils/helpers";
import { EMPTY_FIELD } from "@mds/common/constants/strings";
import { ColumnType } from "antd/lib/table";
import { Button, Dropdown } from "antd";
import { CaretDownOutlined } from "@ant-design/icons";
import { generateActionMenuItems } from "./ActionMenu";

export const renderTextColumn = (
  dataIndex: string,
  title: string,
  sortable = false,
  placeHolder = EMPTY_FIELD,
  width?: number
): ColumnType<any> => {
  return {
    title,
    dataIndex,
    key: dataIndex,
    render: (text: string) => <div title={title}>{text ?? placeHolder}</div>,
    ...(width !== undefined ? { width } : null),
    ...(sortable ? { sorter: nullableStringSorter(dataIndex) } : null),
  };
};

export const renderDateColumn = (
  dataIndex: string,
  title = "Date",
  sortable = false,
  format: (date: any) => string | null = null,
  placeHolder = EMPTY_FIELD
) => {
  const formatFunction = format ?? formatDate;
  return {
    title,
    dataIndex,
    key: dataIndex,
    render: (text) => <div title={title}>{formatFunction(text) || placeHolder}</div>,
    ...(sortable ? { sorter: dateSorter(dataIndex) } : null),
  };
};

export const renderCategoryColumn = (
  dataIndex: string,
  title: string,
  categoryMap: any,
  sortable = false,
  placeHolder = EMPTY_FIELD
) => {
  return {
    title,
    dataIndex,
    key: dataIndex,
    render: (text: string) => <div title={title}>{categoryMap[text] ?? placeHolder}</div>,
    ...(sortable ? { sorter: nullableStringSorter(dataIndex) } : null),
  };
};

export const renderHighlightedTextColumn = (
  dataIndex: string,
  title: string,
  regex: string,
  sortable = true
): ColumnType<any> => {
  return {
    title,
    dataIndex,
    key: dataIndex,
    render: (text: string) => {
      return <Highlight search={regex}>{text}</Highlight>;
    },
    ...(sortable ? { sorter: nullableStringSorter(dataIndex) } : null),
  };
};

export interface ITableAction {
  key: string;
  label: string;
  clickFunction: (event, record) => any;
  icon?: ReactNode;
}

export const renderActionsColumn = (
  actions: ITableAction[],
  recordActionsFilter?: (record, actions) => ITableAction[],
  isRowSelected = false,
  text = "Actions",
  dropdownAltText = "Menu"
) => {
  return {
    key: "actions",
    render: (record) => {
      const filteredActions = recordActionsFilter ? recordActionsFilter(record, actions) : actions;
      const items = generateActionMenuItems(filteredActions, record);

      return (
        <div>
          {items.length > 0 && (
            <Dropdown menu={{ items }} placement="bottomLeft" disabled={isRowSelected}>
              {/* // TODO: change button classname to something generic */}
              <Button data-cy="menu-actions-button" className="permit-table-button">
                {text}
                <CaretDownOutlined alt={dropdownAltText} />
              </Button>
            </Dropdown>
          )}
        </div>
      );
    },
  };
};
