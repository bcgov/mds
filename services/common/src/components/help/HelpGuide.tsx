import { Alert, Button, Col, Drawer, Row, Typography } from "antd";
import React, { FC, useEffect, useState } from "react";
import { Route, Switch, useParams } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";
import { submit } from "redux-form";
import {
  getSystemFlag,
  isAuthenticated,
  userHasRole,
} from "@mds/common/redux/selectors/authenticationSelectors";
import { FORM, SystemFlagEnum, USER_ROLES } from "@mds/common/constants";
import {
  createHelp,
  updateHelp,
  deleteHelp,
  fetchHelp,
  getHelpByKey,
  HELP_GUIDE_ALL_TABS,
  EMPTY_HELP_KEY,
  clearHelp,
} from "@mds/common/redux/slices/helpSlice";
import HelpGuideForm from "./HelpGuideForm";
import { deleteConfirmWrapper } from "../common/ActionMenu";
import { formatSnakeCaseToSentenceCase } from "@mds/common/redux/utils/helpers";
import parse from "html-react-parser";
import { useFeatureFlag } from "@mds/common/providers/featureFlags/useFeatureFlag";
import { Feature } from "@mds/common/utils";
import Loading from "../common/Loading";

interface HelpGuideProps {
  helpKey: string;
}

export const HelpGuideContent: FC<HelpGuideProps> = ({ helpKey }) => {
  const dispatch = useDispatch();
  const system: SystemFlagEnum = useSelector(getSystemFlag);
  const params = useParams<any>();
  const { tab, activeTab } = params;
  const pageTab = tab ?? activeTab;

  const [open, setOpen] = useState(false);
  const [isEditMode, setIsEditMode] = useState(false);
  // it needs to wait for authentication before it can make this check
  const canEditHelp = useSelector((state) => userHasRole(state, USER_ROLES.role_edit_helpdesk));
  const helpGuide = useSelector(getHelpByKey(helpKey, pageTab)) ?? {};
  const { help_guid } = helpGuide;
  const hasHelpGuide = Boolean(help_guid);
  const defaultGuide = helpKey === EMPTY_HELP_KEY;
  const { content = "" } = helpGuide ?? {};
  const [isLoaded, setIsLoaded] = useState(hasHelpGuide);
  const unformattedTitle = pageTab ?? helpKey;
  const title = formatSnakeCaseToSentenceCase(unformattedTitle);
  console.log(helpGuide);

  const showDrawer = () => setOpen(true);
  const hideDrawer = () => {
    setOpen(false);
    setIsEditMode(false);
  };

  const handleFetchData = () => {
    dispatch(fetchHelp({ helpKey, system })).then(() => setIsLoaded(true));
  };

  useEffect(() => {
    dispatch(clearHelp(helpKey));
    handleFetchData();
  }, [helpKey]);

  const handleDeleteGuide = () => {
    deleteConfirmWrapper("Help Guide", () =>
      dispatch(deleteHelp({ helpKey, help_guid })).then(() => {
        setIsEditMode(false);
      })
    );
  };

  const triggerSubmit = () => {
    dispatch(submit(FORM.EDIT_HELP_GUIDE));
  };

  const handleSaveGuide = (values) => {
    const page_tab = values.page_tab ?? HELP_GUIDE_ALL_TABS;
    const data = { ...values, page_tab, system };

    const payload = { helpKey, data };
    const forNewTab = helpGuide.page_tab === "all" && values.page_tab !== "all";
    const saveFunction =
      defaultGuide || forNewTab
        ? () => dispatch(createHelp(payload))
        : () => dispatch(updateHelp(payload));

    saveFunction().then((resp) => {
      if (resp.payload) {
        setIsEditMode(false);
      }
    });
  };

  const getFooterButtons = () => {
    const editButton = (
      <Button type="primary" onClick={() => setIsEditMode(true)}>
        {hasHelpGuide ? "Edit Help Guide" : "Create Help Guide"}
      </Button>
    );
    const submitButton = (
      <Button type="primary" onClick={triggerSubmit}>
        Publish Help Guide
      </Button>
    );
    const deleteButton = (
      <Button disabled={!hasHelpGuide} onClick={handleDeleteGuide}>
        Delete this Guide
      </Button>
    );
    const buttons = isEditMode ? (
      <>
        <Col>{deleteButton}</Col>
        <Col>{submitButton}</Col>
      </>
    ) : (
      <Col>{editButton}</Col>
    );
    const justify = isEditMode ? "space-between" : "end";
    return <Row justify={justify}>{buttons}</Row>;
  };
  const initialValues =
    hasHelpGuide && !defaultGuide ? helpGuide : { help_key: helpKey, page_tab: pageTab };
  const mainContent = isEditMode ? (
    <HelpGuideForm
      initialValues={initialValues}
      handleSaveGuide={handleSaveGuide}
      pageTab={pageTab}
    />
  ) : (
    parse(content)
  );

  return (
    <>
      <Button onClick={showDrawer}>?</Button>
      <Drawer
        placement="right"
        className="help-guide-drawer"
        onClose={hideDrawer}
        open={open}
        title="Help"
        footer={canEditHelp && getFooterButtons()}
        destroyOnClose
      >
        {isEditMode && (
          <Alert
            message="Publish Help Guide"
            showIcon
            type="warning"
            description="Content published here will be visible to MineSpace users. Ensure the language is appropriate and avoid sharing confidential information."
          />
        )}
        <Typography.Title level={2}>{title} Help Guide</Typography.Title>
        {isLoaded ? mainContent : <Loading />}
      </Drawer>
    </>
  );
};

const HelpGuide: FC = () => {
  const dispatch = useDispatch();
  const system: SystemFlagEnum = useSelector(getSystemFlag);
  const authenticated = useSelector(isAuthenticated);
  const { isFeatureEnabled } = useFeatureFlag();
  const helpFeatureEnabled = isFeatureEnabled(Feature.HELP_GUIDE);
  const globalRoutes = Object.entries(GLOBAL_ROUTES).sort(
    (a, b) => (b[1].priority || 0) - (a[1].priority || 0)
  );
  const defaultHelp = useSelector(getHelpByKey(EMPTY_HELP_KEY));

  useEffect(() => {
    if (authenticated && helpFeatureEnabled && !defaultHelp) {
      dispatch(fetchHelp({ helpKey: EMPTY_HELP_KEY, system }));
    }
  }, [authenticated, helpFeatureEnabled]);

  return authenticated && helpFeatureEnabled ? (
    <Switch>
      {globalRoutes.map(([routeName, routeData]) => {
        const { route = "", helpKey = "" } = routeData as any;
        return (
          <Route key={routeName} exact={true} path={route}>
            {helpKey && <HelpGuideContent helpKey={helpKey} />}
          </Route>
        );
      })}
    </Switch>
  ) : null;
};

export default HelpGuide;
