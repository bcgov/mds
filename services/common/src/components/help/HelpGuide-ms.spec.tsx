import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { ReduxWrapper } from "@mds/common/tests/utils/ReduxWrapper";
import HelpGuide, { HelpGuideContent } from "./HelpGuide";
import { AUTHENTICATION } from "@mds/common/constants/reducerTypes";
import { SystemFlagEnum } from "@mds/common/constants";
import { MS_USER_ACCESS_DATA } from "@mds/common/tests/mocks/dataMocks";
import { helpReducerType } from "@mds/common/redux/slices/helpSlice";
import { BrowserRouter } from "react-router-dom";

const msState = {
  [AUTHENTICATION]: {
    systemFlag: SystemFlagEnum.ms,
    userAccessData: MS_USER_ACCESS_DATA,
    isAuthenticated: true,
  },
  [helpReducerType]: {
    helpGuides: {},
  },
};

function mockFunction() {
  const original = jest.requireActual("react-router-dom");
  return {
    ...original,
    useParams: jest.fn().mockReturnValue({
      tab: "overview",
    }),
  };
}

jest.mock("react-router-dom", () => mockFunction());

describe("HelpGuide", () => {
  it("renders MS properly with default content", async () => {
    const helpKey = "Not-Exists";
    const { findByTestId } = render(
      <BrowserRouter>
        <ReduxWrapper initialState={msState}>
          <HelpGuide />
          <HelpGuideContent helpKey={helpKey} />
        </ReduxWrapper>
      </BrowserRouter>
    );
    const helpButton = await findByTestId("help-open");
    fireEvent.click(helpButton);

    const helpContent = await findByTestId("help-content");

    expect(helpContent).toMatchSnapshot();
  });
});
