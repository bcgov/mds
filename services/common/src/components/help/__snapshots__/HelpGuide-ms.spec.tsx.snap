// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpGuide renders MS properly with default content 1`] = `
<div
  data-testid="help-content"
>
  <h2
    class="ant-typography"
    data-testid="help-title"
  >
    Overview
     Help Guide
  </h2>
  <p>
     console.log('should not render as inline function');}&gt;MS default content
  </p>
</div>
`;
